selenium-test/test.js

```diff
@@ -0,0 +1,293 @@
+   reg_url: /www\.youtube\.com/,
+   text: 'yt'
+};
+
+var bangs = [amazon, gimages, bimages, news, wikipedia];
+
+var wd, searchbar, searchbar_text, search_btn;
+
+// Build the webdriver
+function init() {
+        var options = new chrome.Options().addExtensions(EXT_PATH);
+
+        wd = new Builder()
+                .forBrowser('chrome')
+                .setChromeOptions(options)
+                .build();
+}
+
+// Close the browser window
+function tearDown() {
+    wd.quit();
+}
+
+// Test functionality in the popup modal:
+// searchbar, bangs and visible options
+function testSearchbar() {
+        wd.get(POPUP_URL);
+
+        // Test searchbar in the popup modal
+        searchbar = wd.findElement({id:'search_form_input_homepage'});
+        new assert.Assertion(searchbar.isDisplayed(), 'Searchbar exists and is displayed');
+
+        wd.sleep(500);
+
+        searchbar_text = searchbar.getText()
+                .then(function(text){ 
+                new assert.Assertion(text).equals('', 'Searchbar is empty');
+         });
+}
+
+// Test clicking on bangs and then performing a search
+function testBangs(bang) {
+        var bang_btn = wd.findElement({id:BASE_BANG_ID + bang.text});
+    
+    wd.actions().click(bang_btn).perform().then(function(){
+        searchbar = wd.findElement({id:'search_form_input_homepage'});
+        
+        searchbar.getAttribute('value').then(function(text){
+
+                var control_txt = new RegExp('!' + bang.text);
+                var assert_msg = 'Searchbar contains ' + bang.name + ' bang';
+           
+                new assert.Assertion(text).matches(control_txt, assert_msg);
+
+                var search_btn = wd.findElement({id:'search_button_homepage'});
+                var msg = bang.name + ' bang search redirected to the correct site: ';
+                testNewTabUrl(search_btn, msg, bang.reg_url);
+        });
+    });
+}
+
+// Test whether searching using a bang redirects to the correct site
+function testUrl(msg, test_url) {
+   return  wd.getCurrentUrl().then(function(url) {
+        console.log("URL: " + url);
+        new assert.Assertion(url).matches(test_url, msg + url);
+    });
+}
+
+// Test queries in DDG searchbar redirect to duckduckgo.com
+function testDdgSearch() {
+        var x_btn = wd.findElement({id:'search_form_input_clear'});
+        search_btn = wd.findElement({id:'search_button_homepage'});
+
+return  wd.actions()
+        .click(x_btn)
+        .perform()
+        .then(function() {
+                testNewTabUrl(search_btn, 'Searching on DDG', /duckduckgo\.com/);
+        });
+}
+
+// Test newly opened tab and its URL
+function testNewTabUrl(click_el, msg, test_url) {
+        return wd.actions()
+        .click(click_el)
+        .perform()
+        .then(function() {
+                return wd.getAllWindowHandles().then(function(tabs) {
+                        new assert.Assertion(tabs.length).greaterThan(1, 'New tab opened ' + tabs);
```

Comment:
I thought it's a good way to narrow down failures. If the URL matching test fails it might be because it redirected to a wrong URL or because it didn't open a new tab to perform a search at all.

I can skip it if it's a problem tho, sure